### PR TITLE
Add RPC reconnection to the miner

### PR DESCRIPTION
**Pull request by someone235**

### CODE REVIEW

File: cmd/kaspaminer/client.go [hunk 1]
```diff
@@ -5,42 +5,95 @@ import (
...
+
+	retryDuration := time.Second
+	log.Infof("Reconnecting RPC connection")
+	for {
+		err := mc.connect()
+		if err == nil {
+			return
+		}
+
+		const maxRetryDuration = time.Minute
```
Comment (svarogg) [inline cmd/kaspaminer/client.go:49]: I'd rather if this was near the declaration of retryDuration, outside the for

File: cmd/kaspaminer/client.go [hunk 2]
```diff
@@ -5,42 +5,95 @@ import (
...
+		if retryDuration < time.Minute {
+			retryDuration *= 2
+		} else {
+			retryDuration = maxRetryDuration
+		}
+
+		log.Errorf("Got error %s while reconnecting. Trying again in %s", err, retryDuration)
```
Comment (svarogg) [inline cmd/kaspaminer/client.go:56]: Please add single quotes (') around the error, for easier reading.

File: cmd/kaspaminer/client.go [hunk 3]
```diff
@@ -5,42 +5,95 @@ import (
...
-	rpcAddress, err := cfg.NetParams().NormalizeRPCServerAddress(cfg.RPCServer)
-	if err != nil {
-		return nil, err
+func (mc *minerClient) safeRPCClient() *rpcclient.RPCClient {
+	mc.clientLock.RLock()
+	defer mc.clientLock.RUnlock()
+	return mc.rpcClient
+}
+
+func (mc *minerClient) reconnect() {
```
Comment (svarogg) [inline cmd/kaspaminer/client.go:30]: Do we maybe want the reconnect logic on the rpcClient level (rather then minerClient)?

**Review by svarogg — CHANGES_REQUESTED** [review]

File: cmd/kaspaminer/client.go (hunk 3 above)
Comment (someone235) [reply cmd/kaspaminer/client.go:30]: Doing a general purpose reconnection logic is a far more complex task. Maybe we'll do it in the future

File: cmd/kaspaminer/client.go [hunk 4]
```diff
@@ -5,42 +5,95 @@ import (
...
+	cfg                        *configFlags
 	blockAddedNotificationChan chan struct{}
 }
 
-func newMinerClient(cfg *configFlags) (*minerClient, error) {
-	rpcAddress, err := cfg.NetParams().NormalizeRPCServerAddress(cfg.RPCServer)
-	if err != nil {
-		return nil, err
+func (mc *minerClient) safeRPCClient() *rpcclient.RPCClient {
+	mc.clientLock.RLock()
```
[the remark this answers is not in the document]
Comment (someone235) [reply cmd/kaspaminer/client.go:25]: No, it's not atomic, and there's a race condition between this and reconnect

**Review by svarogg — APPROVED** [review]

File: cmd/kaspaminer/client.go (hunk 3 above)
Comment (elichai) [reply cmd/kaspaminer/client.go:30]: Could you open a ticket for that?
IMHO this is pretty critical

**Pull request merged**